src/http/conf.rs

```diff
@@ -1,63 +1,265 @@
+        NonNull::new((*(self.srv_conf).add(module.ctx_index)).cast::<T>())
+    }
+
+    #[inline]
+    unsafe fn http_location_conf_nn_unchecked<T>(&self, module: &ngx_module_t) -> Option<NonNull<T>> {
+        NonNull::new((*(self.loc_conf).add(module.ctx_index)).cast::<T>())
+    }
+}
+
+impl HttpModuleConfExt for ngx_http_upstream_srv_conf_t {
+    #[inline]
+    unsafe fn http_server_conf_nn_unchecked<T>(&self, module: &ngx_module_t) -> Option<NonNull<T>> {
+        let conf = self.srv_conf;
+        if conf.is_null() {
+            return None;
+        }
+        NonNull::new((*conf.add(module.ctx_index)).cast::<T>())
+    }
+}
+
+/// Trait to define and access main module configuration
+pub trait HttpModuleMainConf: HttpModule {
+    /// Type for main module configuration
+    type MainConf;
+    /// Get reference to main module configuration
+    fn main_conf(o: &impl HttpModuleConfExt) -> Option<&'static Self::MainConf> {
+        unsafe { Some(o.http_main_conf_nn_unchecked(Self::module())?.as_ref()) }
+    }
+    /// Get mutable reference to main module configuration
+    fn main_conf_mut(o: &impl HttpModuleConfExt) -> Option<&'static mut Self::MainConf> {
+        unsafe { Some(o.http_main_conf_nn_unchecked(Self::module())?.as_mut()) }
+    }
+}
+
+/// Trait to define and access server-specific module configuration
+pub trait HttpModuleServerConf: HttpModule {
+    /// Type for server-specific module configuration
+    type ServerConf;
+    /// Get reference to server-specific module configuration
+    fn server_conf(o: &impl HttpModuleConfExt) -> Option<&'static Self::ServerConf> {
+        unsafe { Some(o.http_server_conf_nn_unchecked(Self::module())?.as_ref()) }
+    }
+    /// Get mutable reference to server-specific module configuration
+    fn server_conf_mut(o: &impl HttpModuleConfExt) -> Option<&'static mut Self::ServerConf> {
+        unsafe { Some(o.http_server_conf_nn_unchecked(Self::module())?.as_mut()) }
+    }
 }
+
+/// Trait to define and access location-specific module configuration
+pub trait HttpModuleLocationConf: HttpModule {
+    /// Type for location-specific module configuration
+    type LocationConf;
+    /// Get reference to location-specific module configuration
+    fn location_conf(o: &impl HttpModuleConfExt) -> Option<&'static Self::LocationConf> {
+        unsafe { Some(o.http_location_conf_nn_unchecked(Self::module())?.as_ref()) }
+    }
+    /// Get mutable reference to location-specific module configuration
+    fn location_conf_mut(o: &impl HttpModuleConfExt) -> Option<&'static mut Self::LocationConf> {
+        unsafe { Some(o.http_location_conf_nn_unchecked(Self::module())?.as_mut()) }
+    }
+}
+
+mod core {
+    use crate::ffi::{
+        ngx_http_core_loc_conf_t, ngx_http_core_main_conf_t, ngx_http_core_module, ngx_http_core_srv_conf_t,
+    };
+
+    /// Auxiliary structure to access `ngx_http_core_module` module configuration
```

Comment:
The repetition of "module" seems excessive.
```suggestion
    /// Auxiliary structure to access `ngx_http_core_module` configuration.
```